canonicalwebteam/upload_assets/uploader.py

```diff
@@ -42,46 +42,37 @@ def _upload_file(
         }
     )
 
-    if error_on_exists or response.status_code != 409:
-        try:
-            response.raise_for_status()
-        except requests.exceptions.HTTPError as request_error:
-            if request_error.response.status_code == 409:
-                if error_on_exists:
-                    print(
-                        "Error: URL path already exists: {url}".format(
-                            url=os.path.join(
-                                api_url,
-                                url_path
-                            )
-                        ),
-                        file=sys.stderr
-                    )
-                    sys.exit(1)
-                else:
-                    pass
-            elif request_error.response.status_code == 403:
-                print(
-                    (
-                        "Error: Permission denied. "
-                        "Did you provide a valid API token?"
-                    ),
-                    file=sys.stderr
-                )
-                sys.exit(1)
-            elif request_error.response.status_code == 502:
-                print(
-                    (
-                        "Error: Bad gateway. Check the API endpoint, "
-                        "e.g. 'https://assets.ubuntu.com/v1'"
-                    ),
-                    file=sys.stderr
-                )
-                sys.exit(1)
-            else:
-                raise request_error
+    try:
+        response.raise_for_status()
+        asset_info = response.json()
+    except (
+        requests.exceptions.HTTPError,
+        requests.exceptions.JSONDecodeError,
+    ) as request_error:
+        if request_error.response.status_code == 409:
+            print(
+                f"Error: URL path already exists: {url_path}",
+                file=sys.stderr,
+            )
+            sys.exit(1)
+        elif request_error.response.status_code == 403:
+            print(
+                ("Error: Permission denied. " "Did you provide a valid API token?"),
+                file=sys.stderr,
+            )
```

Comment:
I was just curious why do this way and not make request in try and except and then catch error?
I tried a case where I re-uploaded same asset with different name, it gave me this error
![Uploading image.png…]()

Reply:
We're still making the request in the `try-except` block, just that we now also catch `JSONDecodeErrors` in the same block. 
The `raise_for_status` produces predictable status codes, so I'll have to add a branch for 500 errors. In this case, I'll have to debug further because uploading twice even without this tool should return a 429 error - the 500 suggests something went wrong upstream